[2.10] irmin: add extenders in streamed proofs (ter)

samoht opened the pull request:
Co-authored-by: Nicolas Goguey <ngoguey@student.42.fr>
Co-authored-by: Ioana Cristescu <icristescu@users.noreply.github.com>

samoht commented:
A new tentative with a different encoding of extenders (that'll apply in less cases, but we don't really care)

Ngoguey42 commented on src/irmin/proof.ml:
```diff
@@ -132,9 +149,66 @@ struct
+            let i, h = extenders ([ i ], h) ls in
+            match i with
+            | [] -> assert false
+            | [ _ ] -> inode
```
I think this case should be `assert false`

Ngoguey42 replied:
I've changed this

Ngoguey42 reviewed (commented):
I prefer that new solution

samoht commented:
Rebased on top of #1692 ; the tests are failing, I haven't investigated why yet. Also need to re-add the tests that I removed and take the previous reviews into account.

Ngoguey42 commented on src/irmin/tree.ml:
```diff
@@ -1901,7 +1901,15 @@ module Make (P : Private.S) = struct
+      match proofs with
+      | [ (index, proof) ] ->
+          proof_inode_of_node_proof node proof (function
+            | Inode_tree { length; proofs = [ (i, p) ] } ->
+                k (Extender { length; segments = [ index; i ]; proof = p })
```
Disregard that comment

Ngoguey42 commented on src/irmin/tree.ml:
```diff
@@ -1913,7 +1921,18 @@ module Make (P : Private.S) = struct
-      aux [] proofs
+      match proofs with
+      | [ (index, proof) ] ->
+          proof_inode_of_node_proof node proof (function
+            | Inode_tree { length; proofs = [ (i, p) ] } ->
+                k
+                  (Inode_extender { length; segments = [ index; i ]; proof = p })
```
Disregard that comment

Ngoguey42 commented on src/irmin/tree.ml:
```diff
@@ -1901,7 +1901,15 @@ module Make (P : Private.S) = struct
+          proof_inode_of_node_proof node proof (function
+            | Inode_tree { length; proofs = [ (i, p) ] } ->
+                k (Extender { length; segments = [ index; i ]; proof = p })
+            | Inode_extender { length; segments; proof } ->
+                k (Extender { length; segments = index :: segments; proof })
```
I've added this test

icristescu commented on src/irmin/tree.ml:
```diff
@@ -1864,91 +1870,59 @@ module Make (P : Private.S) = struct
       | [ (index, proof) ] ->
-          proof_inode_of_node_proof node proof (function
-            | Inode_tree { length; proofs = [ (i, p) ] } ->
+          (* A is of length 1. *)
```
`length` here is the length of `proofs`, not of the inode, I think the comment is a bit confusing

samoht commented:
LGTM could you clean up the history and then merge?
